Cannot run program "docker-credential-ecr-login.cmd"

**Environment**: git-bash

Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Java version: 1.8.0_252, vendor: Azul Systems, Inc., runtime: D:\Dev\Java\jdk8x64\jre
Default locale: de_DE, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

- *Jib version:* 2.4.0
- *Build tool:* maven 3.6.3
- *OS:* windows

**Description of the issue**:

I installed "docker-credential-ecr-login" ... and I can use it :)

MINGW64 /d/Dev/git/coc/spice-configuration-cluster (develop)
$ docker-credential-ecr-login list                                            
{"https://XXXXXXXXX.dkr.ecr.eu-west-1.amazonaws.com":"AWS"}                
                                                                              
MINGW64 /d/Dev/git/coc/spice-configuration-cluster (develop)
$ docker-credential-ecr-login version                                         
0.6.3                                                                         
                                                                              
MINGW64 /d/Dev/git/coc/spice-configuration-cluster (develop)
$ which docker-credential-ecr-login                                           
/c/Users/hoehmann/go/bin/docker-credential-ecr-login                          

But inside the jib-run it doesn't work.
It seems that jib is trying to execute docker-credential-ecr-login.cmd
instead of docker-credential-ecr-login.exe or docker-credential-ecr-login

**Expected behavior**:

docker-credential-ecr-login should work under windows-git-bash as well

**Steps to reproduce**:

  1.  installed docker-credential-ecr-login with "go get -u github.com/awslabs/amazon-ecr-credential-helper/ecr-login/cli/docker-credential-ecr-login"
  2.  add jib-maven-plugin
  3. run maven

**`jib-maven-plugin` Configuration**: <!-- Delete this section if not used -->
```xml
<configuration>
            <to>
              <image>${aws.ecr.url}/${finalName}:latest</image>
              <credHelper>ecr-login</credHelper>
            </to>
            <container>
              <creationTime>USE_CURRENT_TIMESTAMP</creationTime>
              <ports>
                <port>8080</port>
              </ports>
            </container>
          </configuration>
```

**Log output**: <!-- If applicable, provide relevant log output -->

[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.4.0:build (default-cli) on project rest-service-sce: java.io.IOException: Cannot run program "docker-credential-ecr-login.cmd": CreateProcess error=2, Das System kann die angegebene Datei nicht finden -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.4.0:build (default-cli) on project rest-service-sce: java.io.IOException: Cannot run program "docker-credential-ecr-login.cmd": CreateProcess error=2, Das System kann die angegebene Datei nicht finden
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoExecutionException: java.io.IOException: Cannot run program "docker-credential-ecr-login.cmd": CreateProcess error=2, Das System kann die angegebene Datei nicht finden
    at com.google.cloud.tools.jib.maven.BuildImageMojo.execute (BuildImageMojo.java:162)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: com.google.cloud.tools.jib.registry.credentials.CredentialRetrievalException: java.io.IOException: Cannot run program "docker-credential-ecr-login.cmd": CreateProcess error=2, Das System kann die angegebene Datei nicht finden
    at com.google.cloud.tools.jib.frontend.CredentialRetrieverFactory.lambda$dockerCredentialHelper$1 (CredentialRetrieverFactory.java:150)
    at com.google.cloud.tools.jib.builder.steps.RegistryCredentialRetriever.retrieve (RegistryCredentialRetriever.java:47)
    at com.google.cloud.tools.jib.builder.steps.RegistryCredentialRetriever.getTargetImageCredential (RegistryCredentialRetriever.java:40)
    at com.google.cloud.tools.jib.builder.steps.AuthenticatePushStep.call (AuthenticatePushStep.java:57)
    at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$authenticateBearerPush$3 (StepsRunner.java:236)
    at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly (TrustedListenableFutureTask.java:125)
    at com.google.common.util.concurrent.InterruptibleTask.run (InterruptibleTask.java:69)
    at com.google.common.util.concurrent.TrustedListenableFutureTask.run (TrustedListenableFutureTask.java:78)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
Caused by: java.io.IOException: Cannot run program "docker-credential-ecr-login.cmd": CreateProcess error=2, Das System kann die angegebene Datei nicht finden
    at java.lang.ProcessBuilder.start (ProcessBuilder.java:1048)
    at com.google.cloud.tools.jib.registry.credentials.DockerCredentialHelper.retrieve (DockerCredentialHelper.java:132)
    at com.google.cloud.tools.jib.registry.credentials.DockerCredentialHelper.retrieve (DockerCredentialHelper.java:117)
    at com.google.cloud.tools.jib.frontend.CredentialRetrieverFactory.retrieveFromDockerCredentialHelper (CredentialRetrieverFactory.java:296)
    at com.google.cloud.tools.jib.frontend.CredentialRetrieverFactory.lambda$dockerCredentialHelper$1 (CredentialRetrieverFactory.java:141)
    at com.google.cloud.tools.jib.builder.steps.RegistryCredentialRetriever.retrieve (RegistryCredentialRetriever.java:47)
    at com.google.cloud.tools.jib.builder.steps.RegistryCredentialRetriever.getTargetImageCredential (RegistryCredentialRetriever.java:40)
    at com.google.cloud.tools.jib.builder.steps.AuthenticatePushStep.call (AuthenticatePushStep.java:57)
    at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$authenticateBearerPush$3 (StepsRunner.java:236)
    at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly (TrustedListenableFutureTask.java:125)
    at com.google.common.util.concurrent.InterruptibleTask.run (InterruptibleTask.java:69)
    at com.google.common.util.concurrent.TrustedListenableFutureTask.run (TrustedListenableFutureTask.java:78)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
Caused by: java.io.IOException: CreateProcess error=2, Das System kann die angegebene Datei nicht finden
    at java.lang.ProcessImpl.create (Native Method)
    at java.lang.ProcessImpl.<init> (ProcessImpl.java:444)
    at java.lang.ProcessImpl.start (ProcessImpl.java:139)
    at java.lang.ProcessBuilder.start (ProcessBuilder.java:1029)
    at com.google.cloud.tools.jib.registry.credentials.DockerCredentialHelper.retrieve (DockerCredentialHelper.java:132)
    at com.google.cloud.tools.jib.registry.credentials.DockerCredentialHelper.retrieve (DockerCredentialHelper.java:117)
    at com.google.cloud.tools.jib.frontend.CredentialRetrieverFactory.retrieveFromDockerCredentialHelper (CredentialRetrieverFactory.java:296)
    at com.google.cloud.tools.jib.frontend.CredentialRetrieverFactory.lambda$dockerCredentialHelper$1 (CredentialRetrieverFactory.java:141)
    at com.google.cloud.tools.jib.builder.steps.RegistryCredentialRetriever.retrieve (RegistryCredentialRetriever.java:47)
    at com.google.cloud.tools.jib.builder.steps.RegistryCredentialRetriever.getTargetImageCredential (RegistryCredentialRetriever.java:40)
    at com.google.cloud.tools.jib.builder.steps.AuthenticatePushStep.call (AuthenticatePushStep.java:57)
    at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$authenticateBearerPush$3 (StepsRunner.java:236)
    at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly (TrustedListenableFutureTask.java:125)
    at com.google.common.util.concurrent.InterruptibleTask.run (InterruptibleTask.java:69)
    at com.google.common.util.concurrent.TrustedListenableFutureTask.run (TrustedListenableFutureTask.java:78)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)

**Additional Information**: <!-- Any additional information that may be helpful -->
mvn -DskipTests -Denforcer.skip -nsu jib:build -X

<!-- Thanks for contributing! -->

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java
@@ -118,6 +118,7 @@ public class DockerCredentialHelper {
       try {
         return retrieve(Arrays.asList(credentialHelper.toString() + suffix, "get"));
       } catch (CredentialHelperNotFoundException ignored) {
+        // ignored
       }
     }
     // On Windows, launching a process from Java without a file extension should normally fail, but

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java
@@ -113,7 +113,7 @@ public class DockerCredentialHelper {
       return retrieve(Arrays.asList(credentialHelper.toString(), "get"));
     }
 
-    // We are on Windows.
+    // We are on Windows with undefined/unknown file extension.
     for (String suffix : Arrays.asList(".cmd", ".exe")) {
       try {
         return retrieve(Arrays.asList(credentialHelper.toString() + suffix, "get"));
@@ -121,7 +121,8 @@ public class DockerCredentialHelper {
         // ignored
       }
     }
-    // On Windows, launching a process from Java without a file extension should normally fail, but
+    // On Windows, launching a process from Java without a file extension should normally fail
+    // (https://github.com/GoogleContainerTools/jib/issues/2399#issuecomment-612972912), but
     // running Jib on Linux-like environment (e.g., Cygwin) might succeed?
     return retrieve(Arrays.asList(credentialHelper.toString(), "get"));
   }

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelper.java
@@ -113,16 +113,16 @@ public class DockerCredentialHelper {
       return retrieve(Arrays.asList(credentialHelper.toString(), "get"));
     }
 
-    try {
-      return retrieve(Arrays.asList(credentialHelper.toString() + ".cmd", "get"));
-
-    } catch (CredentialHelperNotFoundException ex) {
+    // We are on Windows.
+    for (String suffix : Arrays.asList(".cmd", ".exe")) {
       try {
-        return retrieve(Arrays.asList(credentialHelper.toString(), "get"));
+        return retrieve(Arrays.asList(credentialHelper.toString() + suffix, "get"));
       } catch (CredentialHelperNotFoundException ignored) {
-        throw ex;
       }
     }
+    // On Windows, launching a process from Java without a file extension should normally fail, but
+    // running Jib on Linux-like environment (e.g., Cygwin) might succeed?
+    return retrieve(Arrays.asList(credentialHelper.toString(), "get"));
   }
 
   private Credential retrieve(List<String> credentialHelperCommand)
@@ -177,7 +177,8 @@ public class DockerCredentialHelper {
 
       // Checks if the failure is due to a nonexistent credential helper CLI.
       if (ex.getMessage().contains("No such file or directory")
-          || ex.getMessage().contains("cannot find the file")) {
+          || ex.getMessage().contains("cannot find the file")
+          || ex.getMessage().contains("error=2")) /* errno=2 (ENOENT) */ {
         throw new CredentialHelperNotFoundException(credentialHelper, ex);
       }
 

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelperTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/credentials/DockerCredentialHelperTest.java
@@ -154,9 +154,11 @@ public class DockerCredentialHelperTest {
       throws CredentialHelperUnhandledServerUrlException, CredentialHelperNotFoundException,
           IOException {
     systemProperties.setProperty("os.name", "WINdows");
-    List<String> errorCommand = Arrays.asList(Paths.get("/foo/bar.cmd").toString(), "get");
+    List<String> errorCmdCommand = Arrays.asList(Paths.get("/foo/bar.cmd").toString(), "get");
+    List<String> errorExeCommand = Arrays.asList(Paths.get("/foo/bar.exe").toString(), "get");
     List<String> command = Arrays.asList(Paths.get("/foo/bar").toString(), "get");
-    Mockito.when(processBuilderFactory.apply(errorCommand)).thenReturn(errorProcessBuilder);
+    Mockito.when(processBuilderFactory.apply(errorCmdCommand)).thenReturn(errorProcessBuilder);
+    Mockito.when(processBuilderFactory.apply(errorExeCommand)).thenReturn(errorProcessBuilder);
     Mockito.when(processBuilderFactory.apply(command)).thenReturn(processBuilder);
 
     DockerCredentialHelper credentialHelper =
@@ -166,7 +168,28 @@ public class DockerCredentialHelperTest {
     Assert.assertEquals("myusername", credential.getUsername());
     Assert.assertEquals("mysecret", credential.getPassword());
 
-    Mockito.verify(processBuilderFactory).apply(errorCommand);
+    Mockito.verify(processBuilderFactory).apply(errorCmdCommand);
+    Mockito.verify(processBuilderFactory).apply(errorExeCommand);
     Mockito.verify(processBuilderFactory).apply(command);
+  }
+
+  @Test
+  public void testRetrieve_fileNotFoundExceptionMessage()
+      throws CredentialHelperUnhandledServerUrlException, IOException {
+    Mockito.when(processBuilderFactory.apply(Mockito.any())).thenReturn(processBuilder);
+    Mockito.when(processBuilder.start())
+        .thenThrow(
+            new IOException(
+                "CreateProcess error=2, Das System kann die angegebene Datei nicht finden"));
+
+    DockerCredentialHelper credentialHelper =
+        new DockerCredentialHelper(
+            "serverUrl", Paths.get("/ignored"), systemProperties, processBuilderFactory);
+    try {
+      credentialHelper.retrieve();
+      Assert.fail();
+    } catch (CredentialHelperNotFoundException ex) {
+      Assert.assertNotNull(ex.getMessage());
+    }
   }
 }

--- a/jib-gradle-plugin/CHANGELOG.md
+++ b/jib-gradle-plugin/CHANGELOG.md
@@ -5,9 +5,13 @@ All notable changes to this project will be documented in this file.
 
 ### Added
 
+- Also tries `.exe` file extension for credential helpers on Windows. ([#2527](https://github.com/GoogleContainerTools/jib/issues/2527))
+
 ### Changed
 
 ### Fixed
+
+- Fixed reporting a wrong credential helper name when the helper does not exist on Windows. ([#2527](https://github.com/GoogleContainerTools/jib/issues/2527))
 
 ## 2.4.0
 
@@ -44,6 +48,7 @@ All notable changes to this project will be documented in this file.
 - Glob pattern support for `jib.extraDirectories.permissions`. ([#1200](https://github.com/GoogleContainerTools/jib/issues/1200))
 - Support for image references with both a tag and a digest. ([#1481](https://github.com/GoogleContainerTools/jib/issues/1481))
 - The `DOCKER_CONFIG` environment variable specifying the directory containing docker configs is now checked during credential retrieval. ([#1618](https://github.com/GoogleContainerTools/jib/issues/1618))
+- Also tries `.cmd` file extension for credential helpers on Windows. ([#2399](https://github.com/GoogleContainerTools/jib/issues/2399))
 
 ### Changed
 

--- a/jib-maven-plugin/CHANGELOG.md
+++ b/jib-maven-plugin/CHANGELOG.md
@@ -5,9 +5,13 @@ All notable changes to this project will be documented in this file.
 
 ### Added
 
+- Also tries `.exe` file extension for credential helpers on Windows. ([#2527](https://github.com/GoogleContainerTools/jib/issues/2527))
+
 ### Changed
 
 ### Fixed
+
+- Fixed reporting a wrong credential helper name when the helper does not exist on Windows. ([#2527](https://github.com/GoogleContainerTools/jib/issues/2527))
 
 ## 2.4.0
 
@@ -44,6 +48,7 @@ All notable changes to this project will be documented in this file.
 - Glob pattern support for `<extraDirectories><permissions>`. ([#1200](https://github.com/GoogleContainerTools/jib/issues/1200))
 - Support for image references with both a tag and a digest. ([#1481](https://github.com/GoogleContainerTools/jib/issues/1481))
 - The `DOCKER_CONFIG` environment variable specifying the directory containing docker configs is now checked during credential retrieval. ([#1618](https://github.com/GoogleContainerTools/jib/issues/1618))
+- Also tries `.cmd` file extension for credential helpers on Windows. ([#2399](https://github.com/GoogleContainerTools/jib/issues/2399))
 
 ### Changed
 

--- a/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/DefaultCredentialRetrievers.java
+++ b/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/DefaultCredentialRetrievers.java
@@ -154,8 +154,8 @@ public class DefaultCredentialRetrievers {
         if (!Files.exists(Paths.get(credentialHelper))) {
           String osName = systemProperties.getProperty("os.name").toLowerCase(Locale.ENGLISH);
           if (!osName.contains("windows")
-              || !Files.exists(Paths.get(credentialHelper + ".cmd"))
-                  && !Files.exists(Paths.get(credentialHelper + ".exe"))) {
+              || (!Files.exists(Paths.get(credentialHelper + ".cmd"))
+                  && !Files.exists(Paths.get(credentialHelper + ".exe")))) {
             throw new FileNotFoundException(
                 "Specified credential helper was not found: " + credentialHelper);
           }

--- a/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/DefaultCredentialRetrievers.java
+++ b/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/DefaultCredentialRetrievers.java
@@ -153,7 +153,9 @@ public class DefaultCredentialRetrievers {
       if (credentialHelper.contains(FileSystems.getDefault().getSeparator())) {
         if (!Files.exists(Paths.get(credentialHelper))) {
           String osName = systemProperties.getProperty("os.name").toLowerCase(Locale.ENGLISH);
-          if (!osName.contains("windows") || !Files.exists(Paths.get(credentialHelper + ".cmd"))) {
+          if (!osName.contains("windows")
+              || !Files.exists(Paths.get(credentialHelper + ".cmd"))
+                  && !Files.exists(Paths.get(credentialHelper + ".exe"))) {
             throw new FileNotFoundException(
                 "Specified credential helper was not found: " + credentialHelper);
           }

--- a/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/DefaultCredentialRetrieversTest.java
+++ b/jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/DefaultCredentialRetrieversTest.java
@@ -270,7 +270,7 @@ public class DefaultCredentialRetrieversTest {
   }
 
   @Test
-  public void testCredentialHelper() throws IOException {
+  public void testCredentialHelper_cmdExtension() throws IOException {
     Path credHelper = temporaryFolder.newFile("foo.cmd").toPath();
     Path pathWithoutCmd = credHelper.getParent().resolve("foo");
     Assert.assertEquals(pathWithoutCmd.getParent().resolve("foo.cmd"), credHelper);
@@ -291,6 +291,47 @@ public class DefaultCredentialRetrieversTest {
     List<CredentialRetriever> retrieverList =
         new DefaultCredentialRetrievers(mockCredentialRetrieverFactory, properties, environment)
             .setCredentialHelper(pathWithoutCmd.toString())
+            .asList();
+
+    Assert.assertEquals(
+        Arrays.asList(
+            mockDockerCredentialHelperCredentialRetriever,
+            mockDockerConfigEnvDockerConfigCredentialRetriever,
+            mockDockerConfigEnvKubernetesDockerConfigCredentialRetriever,
+            mockDockerConfigEnvLegacyDockerConfigCredentialRetriever,
+            mockSystemHomeDockerConfigCredentialRetriever,
+            mockSystemHomeKubernetesDockerConfigCredentialRetriever,
+            mockSystemHomeLegacyDockerConfigCredentialRetriever,
+            mockEnvHomeDockerConfigCredentialRetriever,
+            mockEnvHomeKubernetesDockerConfigCredentialRetriever,
+            mockEnvHomeLegacyDockerConfigCredentialRetriever,
+            mockWellKnownCredentialHelpersCredentialRetriever,
+            mockApplicationDefaultCredentialRetriever),
+        retrieverList);
+  }
+
+  @Test
+  public void testCredentialHelper_exeExtension() throws IOException {
+    Path credHelper = temporaryFolder.newFile("foo.exe").toPath();
+    Path pathWithoutExe = credHelper.getParent().resolve("foo");
+    Assert.assertEquals(pathWithoutExe.getParent().resolve("foo.exe"), credHelper);
+
+    DefaultCredentialRetrievers credentialRetrievers =
+        new DefaultCredentialRetrievers(mockCredentialRetrieverFactory, properties, environment)
+            .setCredentialHelper(pathWithoutExe.toString());
+    try {
+      credentialRetrievers.asList();
+      Assert.fail("shouldn't check .exe suffix on non-Windows");
+    } catch (FileNotFoundException ex) {
+      Assert.assertThat(
+          ex.getMessage(), CoreMatchers.startsWith("Specified credential helper was not found:"));
+      Assert.assertThat(ex.getMessage(), CoreMatchers.endsWith("foo"));
+    }
+
+    properties.setProperty("os.name", "winDOWs");
+    List<CredentialRetriever> retrieverList =
+        new DefaultCredentialRetrievers(mockCredentialRetrieverFactory, properties, environment)
+            .setCredentialHelper(pathWithoutExe.toString())
             .asList();
 
     Assert.assertEquals(

